src/routes/conversation/[id]/+server.ts

```diff
@@ -410,6 +429,173 @@ export async function POST({ request, locals, params, getClientAddress }) {
 				}
 			}
 
+			const endpoint = await model.getEndpoint();
+			let toolResults: ToolResult[] | undefined = undefined;
+
+			// function calls, if no assistant or assistant has websearch enabled
+			if (model.functions && (!assistant || assistantHasWebSearch)) {
+				let tools: BackendTool[] = [];
+
+				// if it's an assistant, only support websearch for now
+				if (!assistant) {
+					// filter based on tool preferences, add the tools that are on by default
+					tools = [directlyAnswer, text2img, calculator, fetchUrl, websearch].filter((el) => {
+						if (el.isLocked && el.isOnByDefault) {
+							return true;
+						}
+
+						return toolsPreferences?.[el.name] ?? el.isOnByDefault;
+					});
+				} else {
+					// an assistant can currently only use tools for websearch
+					tools = [directlyAnswer, websearch];
+				}
+
+				let calls: Call[] | undefined = undefined;
+
+				// do the function calling bits here
+				for await (const output of await endpoint({
+					messages: messagesForPrompt,
+					preprompt,
+					generateSettings: assistant?.generateSettings,
+					tools,
+				})) {
+					if (output.generated_text) {
+						// look for a code blocks of ```json and parse them
+						// if they're valid json, add them to the calls array
+						const codeBlocks = output.generated_text.match(/```json\n(.*?)```/gs);
+						if (codeBlocks) {
+							for (const block of codeBlocks) {
+								try {
+									calls = JSON5.parse(block.replace("```json\n", "").slice(0, -3));
```

Comment:
in `.slice(0, -3)`, what is `-3`? It would be much better to tie it to a specific variable like `-xyz.length`